### PR TITLE
[CIS-2180] Update CI to leverage Xcode 14.0.1

### PR DESCRIPTION
**Pull request by alteral**
### 🔗 Issue Links

- https://stream-io.atlassian.net/browse/CIS-2180

### 🎯 Goal

- Update CI to leverage Xcode 14.0.1
- Update snapshots
- Make it easy to record new snapshots

### Notes

- Snapshot tests stay on Xcode 13.4.1 due to some issues with snapshots on iOS 16
- I was not able to fix this unit test so I removed it and created [a ticket](https://stream-io.atlassian.net/browse/CIS-2206) as a tech debt to fix it:

```
JSONDecoder_Tests.test_defaultDecoder_isStreamDecoder()
```

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

gif

### CODE REVIEW

File: Sources/StreamChatUI/Gallery/VideoPlaybackControlView.swift [hunk 1]
```diff
@@ -52,6 +52,7 @@ open class VideoPlaybackControlView: _View, ThemeProvider {
     private var playerStatusObserver: NSKeyValueObservation?
     private var playerItemObserver: NSKeyValueObservation?
     private var itemDurationObserver: NSKeyValueObservation?
+    private var _currentPlayer: AVPlayer? { player }
```
Comment (nuno-vieira) [inline Sources/StreamChatUI/Gallery/VideoPlaybackControlView.swift:55]: I would move this closer to the deinit method, and provide a comment to explain this is to overcome a Swift 6 warning 👍

File: Tests/StreamChatUITests/SnapshotTests/ChatMessageList/ChatMessage/ChatMessageContentView_Tests.swift [hunk 2]
```diff
@@ -196,59 +196,55 @@ final class ChatMessageContentView_Tests: XCTestCase {
...
+//          **bold** or __bold__
+//          ~~Linethrough~~Strikethroughs.
+//          `let property: Double = 10.0`
+//
+//          # Header 1
+//          ## Header 2
+//          ### Header 3
+//          #### Header 4
+//          ##### Header 5 #####
+//          ###### Header 6 ######
```
Comment (nuno-vieira) [inline Tests/StreamChatUITests/SnapshotTests/ChatMessageList/ChatMessage/ChatMessageContentView_Tests.swift:215]: Why is this test commented out? 🤔

File: .github/workflows/smoke-checks.yml [hunk 3]
```diff
@@ -22,68 +22,67 @@ concurrency:
 
 env:
   HOMEBREW_NO_INSTALL_CLEANUP: 1 # Disable cleanup for homebrew, we don't need it on CI
-  XCODE_VERSION: 13.4.1
-  IOS_VERSION: "15.5"
-  IOS_SIMULATOR_DEVICE: "iPhone 12 Pro"
-  IOS_SIMULATOR_DEVICE_WITH_VERSION: "iPhone 12 Pro (15.5)"
+  XCODE_VERSION: "14.0.1"
+  IOS_VERSION: "15.4"
+  IOS_SIMULATOR_DEVICE_WITH_VERSION: "iPhone 12 Pro (15.4)"
```
Comment (polqf) [inline .github/workflows/smoke-checks.yml:27]: Why did we downgrade here?

File: .github/workflows/smoke-checks.yml [hunk 4]
```diff
@@ -22,68 +22,67 @@ concurrency:
...
+  #   if: ${{ github.event_name != 'push' }}
+  #   steps:
+  #     - uses: actions/checkout@v2
+  #     - uses: ./.github/actions/bootstrap
+  #     - name: Run Danger
+  #       run: bundle exec danger
+  #       env:
+  #         GITHUB_TOKEN: ${{ secrets.DANGER_GITHUB_API_TOKEN }}
+  #     - name: Run Linting
+  #       run: ./Scripts/run-linter.sh
```
Comment (polqf) [inline .github/workflows/smoke-checks.yml:42]: Are all these flows meant to be disabled?

File: .github/workflows/smoke-checks.yml [hunk 5]
```diff
@@ -92,82 +91,84 @@ jobs:
...
       timeout-minutes: 25
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         GITHUB_EVENT: ${{ toJson(github.event) }}
+    - uses: actions/upload-artifact@v3
+      if: failure()
+      with:
+        name: snapshots
+        path: |
+          Tests/StreamChatUITests/SnapshotTests/ChatChannel/__Snapshots__/ChatChannelVC_Tests/test_whenMessageFromCurrentUserIsFollowedByEphemeralMessage_messagesAreNotGrouped.default-light.png
```
Comment (polqf) [inline .github/workflows/smoke-checks.yml:107]: Is this only for testing purposes?

File: Examples/SlackClone/SlackReactionsMessagePopupVC.swift [hunk 6]
```diff
@@ -7,8 +7,8 @@ import StreamChatUI
 
 final class SlackReactionsMessagePopupVC: ChatMessagePopupVC {
     override func setUpLayout() {
-        let messageView = messageContentView as! SlackChatMessageContentView
-        messageView.isInPopupView = true
+        let messageView = messageContentView as? SlackChatMessageContentView
+        messageView?.isInPopupView = true
```
Comment (polqf) [inline Examples/SlackClone/SlackReactionsMessagePopupVC.swift:11]: 👌

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift [hunk 7]
```diff
@@ -1014,12 +1014,12 @@ extension ChatMessageContentView {
     }
     
     @available(*, deprecated, renamed: "createOnlyVisibleToYouImageView")
-    open func createOnlyVisibleForYouIconImageView() -> UIImageView {
+    public func createOnlyVisibleForYouIconImageView() -> UIImageView {
```
Comment (polqf) [inline Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1017]: Why this change? It can be a breaking change

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift [hunk 8]
```diff
@@ -1014,12 +1014,12 @@ extension ChatMessageContentView {
...
         createOnlyVisibleToYouImageView()
     }
     
     @available(*, deprecated, renamed: "createOnlyVisibleToYouLabel")
-    open func createOnlyVisibleForYouLabel() -> UILabel {
+    public func createOnlyVisibleForYouLabel() -> UILabel {
```
Comment (polqf) [inline Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1022]: Same here

File: Sources/StreamChatUI/Deprecations.swift [hunk 9]
```diff
@@ -290,7 +290,7 @@ extension AttachmentsPreviewVC {
         deprecated,
         message: "it has been replaced by attachmentPreviews(for:). The name reflects better the intent and when asking for the attachment views, it is safer to specify which axis or axises we want."
     )
-    open var attachmentViews: [UIView] {
+    public var attachmentViews: [UIView] {
```
Comment (polqf) [inline Sources/StreamChatUI/Deprecations.swift:293]: Same in this file

File: TestTools/StreamChatTestMockServer/StreamChatTestMockServer.swift [hunk 10]
```diff
@@ -34,6 +33,6 @@ extension Bundle {
     }()
 
     var pathToJSONsFolder: String {
-        Self.testTools.bundlePath.contains(Self.bundleName) ? Self.resourcesJSONs : ""
+        ""
```
Comment (polqf) [inline TestTools/StreamChatTestMockServer/StreamChatTestMockServer.swift:36]: Is this correct?

**Review by polqf — COMMENTED** [review]
Is this PR ready to be reviewed?

**Comment by alteral** [comment]
> Is this PR ready to be reviewed?

Sorry, not really. Still fighting with the snapshot tests. But most of your comments are relevant, I'll answer them as soon as fix those flaky snapshot tests.

File: .github/workflows/smoke-checks.yml (hunk 5 above)
Comment (alteral) [reply .github/workflows/smoke-checks.yml:107]: Yeah, debugging failing snapshot test on CI

File: .github/workflows/smoke-checks.yml (hunk 4 above)
Comment (alteral) [reply .github/workflows/smoke-checks.yml:42]: Nah, just debugging failing snapshot test on CI

File: Sources/StreamChatUI/Gallery/VideoPlaybackControlView.swift (hunk 1 above)
Comment (alteral) [reply Sources/StreamChatUI/Gallery/VideoPlaybackControlView.swift]: Done 👍

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift (hunk 7 above)
Comment (alteral) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1017]: Xcode 14 raised a warning here. If this is not the right solution, please let me know what's the best way to fix it.

File: TestTools/StreamChatTestMockServer/StreamChatTestMockServer.swift (hunk 10 above)
Comment (alteral) [reply TestTools/StreamChatTestMockServer/StreamChatTestMockServer.swift:36]: Yeah, this is the workaround to fix the warning in SwiftUI repo. It might not look great, but it follows the pattern of having  `pathToJSONsFolder` in both `TestTools` and `MockServer`.

File: .github/workflows/smoke-checks.yml (hunk 3 above)
Comment (alteral) [reply .github/workflows/smoke-checks.yml:27]: Egh, I did not find a way how to install `15.5` on Xcode 14. We only have `16` and `15.4` out there. As snapshot tests do not work well enough on iOS `16`, the only option is `15.4`. See [failed job](https://github.com/GetStream/stream-chat-swift/actions/runs/3225236953/jobs/5277303724) and screenshot attached:
Screen Shot 2022-10-11 at 9 24 24 AM

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift (hunk 7 above)
Comment (polqf) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1017]: We need to keep it as open, otherwise our integrators cannot override those methods. We can tackle these warnings later on

Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1017]: The warning is saying that there is no open functions in extensions. So right now this is not working. Or we move this deprecation in the actual file, or we just leave it public. If no one compplained about this yet, it means nobody is using it anyways

**Comment by alteral** [comment]
This is now ready for review, thanks 🚢

File: Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift (hunk 7 above)
Comment (polqf) [reply Sources/StreamChatUI/ChatMessageList/ChatMessage/ChatMessageContentView.swift:1017]: Oh ok! Let's move to public then

File: .github/workflows/smoke-checks.yml [hunk 11]
```diff
@@ -92,18 +88,13 @@ jobs:
...
     - uses: ./.github/actions/bootstrap
-    - name: "Prepare ${{ env.IOS_SIMULATOR_DEVICE }} with iOS ${{ env.IOS_VERSION }}"
-      uses: ./.github/actions/prepare-ios-simulator
-      with:
-        device: ${{ env.IOS_SIMULATOR_DEVICE }}
-        ios: ${{ env.IOS_VERSION }}
-        xcode: ${{ env.XCODE_VERSION }}
     - name: Run UI Tests (Debug)
-      run: bundle exec fastlane test_ui device:"${{ env.IOS_SIMULATOR_DEVICE_WITH_VERSION }}"
+      run: bundle exec fastlane test_ui device:"iPhone 12 Pro (15.5)"
```
Comment (polqf) [inline .github/workflows/smoke-checks.yml:92]: Isn't it better to keep the env variables?

**Review by polqf — COMMENTED** [review]
Any reason why the snapshots went from 1080 to 360 px width?

File: .github/workflows/smoke-checks.yml (hunk 11 above)
Comment (alteral) [reply .github/workflows/smoke-checks.yml:92]: `Any reason why the snapshots went from 1080 to 360 px width?`
- I removed our custom action to prepare the Simulator (which was legacy to save time on CI) and used the default iPhone 12 Pro (15.5.) available on Github Actions.

`Isn't it better to keep the env variables?`
- Those env variables were used only in this job but were set up on the pipeline level. The current implementation aligns with other jobs (unit & e2e).

Comment (alteral) [reply .github/workflows/smoke-checks.yml:92]: Although, on the SwiftUI repo it uses 1080px width with the same setup, weird 🤔

Comment (alteral) [reply .github/workflows/smoke-checks.yml]: Sorry, my bad, returned the `IOS_SIMULATOR_DEVICE` env variable.

File: .github/actions/prepare-ios-simulator/action.yml [hunk 12]
```diff
@@ -1,28 +0,0 @@
...
-
-        sudo mkdir -p /Library/Developer/CoreSimulator/Profiles/Runtimes
-
-        sudo ln -s /Applications/Xcode_$xcode_version.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime /Library/Developer/CoreSimulator/Profiles/Runtimes/iOS\ $ios_version.simruntime
-
-        xcrun simctl list runtimes
-        xcrun simctl create custom-test-device "${{ inputs.device }}" "com.apple.CoreSimulator.SimRuntime.iOS-$ios_version_dash"
-        xcrun simctl list devices $ios_version
-      shell: bash
-
```
Comment (nuno-vieira) [inline .github/actions/prepare-ios-simulator/action.yml:28]: Why we don't need this anymore?

File: fastlane/Fastfile [hunk 13]
```diff
@@ -736,6 +738,7 @@ desc 'Build and upload DemoApp to Emerge'
 lane :emerge_upload do
   # Build and upload Release config
   build_ios_app(
+    export_method: 'ad-hoc',
```
Comment (nuno-vieira) [inline fastlane/Fastfile:741]: Can we remove this? I remember we tested this to fix Emerge, but this was not the issue, so better to revert this

File: fastlane/Fastfile [hunk 14]
```diff
@@ -777,13 +780,18 @@ lane :rubocop do
   sh('bundle exec rubocop')
 end
 
+desc 'Remove UI snapshots'
+private_lane :remove_snapshots do
+  Dir.glob("../Tests/StreamChatUITests/**/__Snapshots__/**/*.png").select { |file| File.delete(file) }
+end
```
Comment (nuno-vieira) [inline fastlane/Fastfile:786]: What is the goal of this? 🤔

**Review by nuno-vieira — COMMENTED** [review]
Overall looks solid 👍 Just added a couple of questions. Also, why all snapshots are now 360px 🤔

File: .github/workflows/smoke-checks.yml (hunk 11 above)
Comment (nuno-vieira) [reply .github/workflows/smoke-checks.yml]: The snapshots are recorded in your machine, not in the CI. So I wonder if this is related to updating the snapshot library or just Xcode 14 🤔 But that would mean that snapshot tests would fail for who ever is still on Xcode 13

Comment (alteral) [reply .github/workflows/smoke-checks.yml]: Nevermind, this is because I bumped `swift-snapshot-testing` lib version. Just reverted it back.

File: fastlane/Fastfile (hunk 14 above)
Comment (alteral) [reply fastlane/Fastfile:786]: To remove all png files if argument `record` is provided. I found it easy to collect snapshots from cli using fastlane.

File: fastlane/Fastfile (hunk 13 above)
Comment (alteral) [reply fastlane/Fastfile:741]: Actually, I think it's good to have it because the error message had much more sense when`export_method` was provided.

File: .github/actions/prepare-ios-simulator/action.yml (hunk 12 above)
Comment (alteral) [reply .github/actions/prepare-ios-simulator/action.yml:28]: It was a legacy to save time on CI to configure the simulator. Right now everything works out of the box on Github Actions.

File: .github/workflows/smoke-checks.yml (hunk 11 above)
Comment (nuno-vieira) [reply .github/workflows/smoke-checks.yml]: Makes sense now 👍 But this is actually interesting, it reduced the git repo size. This is probably an optimization. But lets reduce the scope of this PR for now, and we handle that later 👍

**Review by nuno-vieira — APPROVED** [review]
LGTM! 🚀
